framework/src/media/media_recorder.c

```diff
@@ -0,0 +1,348 @@
+#include <media/media_utils.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <fcntl.h>
+
+#define IS_SOCKET(fd) (fd > CONFIG_NFILE_DESCRIPTORS)
+#define RECORD_LOCK() do { pthread_mutex_lock(&g_rc.mutex_record); } while (false)
+#define RECORD_UNLOCK() do { pthread_mutex_unlock(&g_rc.mutex_record); } while (false)
+
+#define ST_PROTO_RECORD			0x04
+#define ST_PROTO_RECORD_DATA	0x05
+#define ST_PROTO_RECORD_PAUSE	0x02
+#define ST_PROTO_RECORD_STOP	0x03
+
+struct media_record_context_s {
+	pthread_t           pth;
+	bool                thread_running;
+	pthread_mutexattr_t mutex_attr_record;
+	pthread_mutex_t     mutex_record;
+	sem_t               sem_pause;
+	struct pcm         *pcmin;
+	struct pcm_config  *pcmconf;
+	unsigned int        buffer_size;
+	int                 fd;
+	media_format_e      format;
+	record_state_t      state;
+};
+
+static struct media_record_context_s g_rc;
+static bool b_started = false;
+
+int media_recording(void *args);
+
+/**
+ * Public Functions
+ */
+record_result_t media_record(int fd)
+{
+	pthread_attr_t attr;
+	struct sched_param sparam;
+
+	g_rc.fd = fd;
+
+	if (IS_SOCKET(g_rc.fd)) {
+		char instruction = ST_PROTO_RECORD;
+		send(g_rc.fd, &instruction, 1, 0);	
+	}
+	
+	RECORD_LOCK();
+
+	if (g_rc.state != RECORD_STOPPING && g_rc.state != RECORD_AVAILABLE) {
+		RECORD_UNLOCK();
+		return RECORD_ALREADY_INITIALIZED;
+	}
+
+	if (g_rc.state == RECORD_STOPPING) {
+		pthread_join(g_rc.pth, NULL);
+		g_rc.state = RECORD_AVAILABLE;
+	}
+
+	if (g_rc.state == RECORD_AVAILABLE) {
+		g_rc.fd = fd;
+		g_rc.thread_running = true;
+
+		if (pthread_attr_init(&attr) != 0) {
+			goto error_thread_init;
+		}
+
+		sparam.sched_priority = CONFIG_MEDIA_PRIORITY;
+		if (pthread_attr_setschedparam(&attr, &sparam) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setschedpolicy(&attr, SCHED_RR) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setstacksize(&attr, CONFIG_MEDIA_STACK_SIZE) != 0) {
+			goto error_thread_init;
+		}
+		if (pthread_create(&g_rc.pth, &attr, (pthread_startroutine_t)media_recording, NULL) != 0) {
+			goto error_thread_init;
+		}
+
+		g_rc.state = RECORD_RUNNING;
+	}
+
+	RECORD_UNLOCK();
+	return RECORD_OK;
+
+error_thread_init:
+	g_rc.thread_running = false;
+	RECORD_UNLOCK();
+	return RECORD_ERROR_THREAD_INIT;
+}
+
+record_result_t media_record_set_config(int channel, int sample_rate, int pcm_format, media_format_e format)
+{
+	RECORD_LOCK();
+
+	if (g_rc.pcmconf) {
+		RECORD_UNLOCK();
+		return RECORD_ALREADY_INITIALIZED;
+	}
+
+	g_rc.pcmconf = (struct pcm_config *)malloc(sizeof(struct pcm_config));
+	if (!g_rc.pcmconf) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_MALLOC;
+	}
+
+	g_rc.pcmconf->channels = channel;
+	g_rc.pcmconf->rate = sample_rate;
+	g_rc.pcmconf->format = pcm_format;
+	g_rc.format = format;
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_record_prepare(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.pcmin) {
```

Comment:
After finishing recording, we should re-call this to record another. But due to first recording, this variable isn't NULL so that below pcm_open will not be called. Isn't it?

Reply:
should we call init-prepare-record-stop-init-prepare-record? not init-prepare-record-stop-record-stop or not init-prepare-record-stop-prepare-record-stop?

Reply:
@Taejun-Kwon taejun we should discuss these with pcm drop, drain

Reply:
@sunghan-chang next steps, we will change related network and deinit.